### PR TITLE
Fix a bug in 'extract.timestamps' when dealing with an empty data source

### PR DESCRIPTION
**Pull request by maxloeffler**
### Prerequisites

- [x] I adhere to the coding conventions (described [here](../CONTRIBUTING.md)) in my code.
- [x] I have updated the copyright headers of the files I have modified.
- [x] I have written appropriate commit messages, i.e., I have recorded the goal, the need, the needed changes, and the location of my code modifications for each commit. This includes also, e.g., referencing to relevant issues.
- [x] I have put signed-off tags in *all* commits.
- [x] I have updated the changelog file [NEWS.md](../NEWS.md) appropriately.
- [x] I have checked whether I need to adjust the showcase file `showcase.R` with respect to my changes.
- [x] The pull request is opened against the branch `dev`.

### Description

Include a fix by @bockthom for issue #269 and a new unit test for the previously faulty behavior.

### Changelog

Fixed:
* Fix a bug in `extract.timestamps` that occurs when the first `data.source` is empty and leads to a return value of type `numeric` which should be `POSIXct`.

### CODE REVIEW

File: tests/test-data-cut.R [hunk 1]
```diff
@@ -82,3 +83,58 @@ test_that("Cut commit and mail data to same date range.", {
...
+                                        committer.email = c("bjoern@example.org", "bjoern@example.org"),
+                                        hash = c("72c8dd25d3dd6d18f46e2b26a5f5b1e2e8dc28d0", "5a5ec9675e98187e1e92561e1888aa6f04faa338"),
+                                        changed.files = as.integer(c(1, 1)),
+                                        added.lines = as.integer(c(1, 1)),
+                                        deleted.lines = as.integer(c(1, 0)),
+                                        diff.size = as.integer(c(2, 1)),
+                                        file = c("test.c", "test.c"),
+                                        artifact = c("A", "A"),
+                                        artifact.type = c("Feature", "Feature"),
+                                        artifact.diff.size = as.integer(c(1, 1)))
```
Comment (bockthom) [inline tests/test-data-cut.R:115]: Could you please align the indentation with the beginning of the first parameter?

File: tests/test-data-cut.R [hunk 2]
```diff
@@ -82,3 +83,58 @@ test_that("Cut commit and mail data to same date range.", {
...
+    ## configurations
+
+    proj.conf = ProjectConf$new(CF.DATA, CF.SELECTION.PROCESS, CASESTUDY, ARTIFACT)
+    proj.conf$update.value("issues.locked", TRUE)
+    data.sources = c("issues", "mails", "commits")
+
+    ## construct objects
+
+    x.data = ProjectData$new(proj.conf)
+    x.data$set.issues(NULL)
```
Comment (bockthom) [inline tests/test-data-cut.R:98]: Could you please add a comment that states that, in this test, actually the first processed data source is empty, to highlight that the first processed empty determines the data type of the data frame? This is an important aspect that has to be considered, for example, when this test is to be adjusted in the future.

File: tests/test-data-cut.R [hunk 3]
```diff
@@ -82,3 +83,58 @@ test_that("Cut commit and mail data to same date range.", {
...
+    rownames(commit.data) = 1:nrow(commit.data)
+
+    mail.data = x.data$get.data.cut.to.same.date(data.sources = data.sources)$get.mails()
+    rownames(mail.data) = 1:nrow(mail.data)
+
+    issue.data = x.data$get.data.cut.to.same.date(data.sources = data.sources)$get.issues()
+
+    expect_identical(commit.data, commit.data.expected, info = "Cut Raw commit data.")
+    expect_identical(mail.data, mail.data.expected, info = "Cut mail data.")
+    expect_identical(issue.data, issue.data.expected, info = "Cut issue data.")
```
Comment (bockthom) [inline tests/test-data-cut.R:138]: Maybe highlight that this empty: "Cut issue data (empty)."

File: util-data.R [hunk 4]
```diff
@@ -113,7 +113,7 @@ DATASOURCE.TO.ARTIFACT.COLUMN = list(
 
 
 ## the maximum time difference between subsequent mails of a patchstack
-PATCHSTACK.MAIL.DECAY.THRESHOLD = "30 seconds"
+PATCHSTACK.MAIL.DECAY.THRESHOLD = lubridate::as.duration("30 seconds")
```
Comment (bockthom) [inline util-data.R:116]: Please update the copyright header in this file.

**Review by bockthom — COMMENTED** [review]
Thanks for this PR @MaLoefUDS.  I had a quick look at it, looks good to me. I just spotted a few minor issues, such as enhancing comments with additional information.

And it also seems that you need to rebase onto the current `dev` branch.

File: tests/test-data-cut.R (hunk 2 above)
Comment (MaLoefUDS) [reply tests/test-data-cut.R:98]: Im struggeling a bit with formulating this comment. My problem is that now, after your fix, the first data source does not influence the data type of the dataframe anymore, as this was never intended. Currently, I'm stuck with something like

```
    ## ensure that the first data source contains empty data, as it used to determine
    ## the data type of the data frame that stores the project timestamps (which is unintended)
```

Comment (bockthom) [reply tests/test-data-cut.R:98]: That sounds already good to me. I just would like to highlight even more that this is important for the test. 
And of course it is intended that the first entry determines the data type of the data frame, this is what is going on in R. This is why it is important that the first entry has the correct data type.

So, what about the following:

```R
## in order to test whether the data types of timestamps originating from empty data are correct, 
## ensure that the first data source contains empty data, as it is used to determine
## the data type of the data frame that stores the project timestamps
```

Comment (MaLoefUDS) [reply tests/test-data-cut.R:98]: > And of course it is intended that the first entry determines the data type of the data frame, this is what is going on in R.

>  `as it is used to determine the data type of the data frame that stores the project timestamps`

I have to disagree with you on this one. While on a language level this is the intended behavior, it is not for the output of the `extract.timestamps` method and the latter propagation of this value in other coronet components. This is especially highlighted by the fact that after your fix the data type of the project timestamps data frame is totally uniform.

I find it challenging to find a fitting comment that states that the position of the empty data source mattered before your fix without explicitly mentioning your fix.

Comment (bockthom) [reply tests/test-data-cut.R:98]: Ok, your concerns are valid.

What about the following wording:

```R
... as R usually uses the first entry of a data frame to determine the data type of the data frame but this should be of no effect since the data type of the project timestamp should be properly set anyway.
```
What do you think about it?

I guess we need to split this up into multiple sentences, as the first part of this sentence is already quite long.

Comment (MaLoefUDS) [reply tests/test-data-cut.R:98]: Sounds good. Let me see what I can come up with

**Comment by MaLoefUDS** [comment]
*The last force push was just because of rebasing

**Comment by bockthom** [comment]
> *The last force push was just because of rebasing

That's what I have thought anyway :smile:

**Review by bockthom — DISMISSED** [review]

**Review by bockthom — APPROVED** [review]
